embedded-graphics/src/draw_target.rs

```diff
@@ -0,0 +1,322 @@
+///     framebuffer: [0; 4096],
+///     iface: SPI1,
+/// };
+///
+/// // Draw a rectangle from (10, 20) to (30, 40) with a white stroke
+/// let rect = egrectangle!(
+///     top_left = (10, 20),
+///     bottom_right = (30, 40),
+///     style = primitive_style!(stroke_color = Gray8::WHITE, stroke_width = 1)
+/// );
+/// rect.draw(&mut display)?; // Uses the accelerated draw_rectangle function
+///
+/// // Update the display
+/// display.flush().expect("Failed to send data to display");
+/// # Ok::<(), core::convert::Infallible>(())
+/// ```
+///
+/// [`Drawable`]: ../drawable/trait.Drawable.html
+/// [`Pixel`]: ../drawable/struct.Pixel.html
+/// [`draw_pixel`]: ./trait.DrawTarget.html#method.draw_pixel
+/// [`DrawTarget`]: ./trait.DrawTarget.html
+/// [`Rectangle`]: ../primitives/rectangle/struct.Rectangle.html
+pub trait DrawTarget<C>
+where
+    C: PixelColor,
+{
+    /// Error type to return when a drawing operation fails.
+    ///
+    /// This error is returned if an error occurred during a drawing operation. This mainly applies
+    /// to drivers that need to communicate with the display for each drawing operation, where a
+    /// communication error can occur. For drivers that use an internal framebuffer where drawing
+    /// operations can never fail, [`core::convert::Infallible`] can instead be used as the `Error`
+    /// type.
+    ///
+    /// [`core::convert::Infallible`]: https://doc.rust-lang.org/stable/core/convert/enum.Infallible.html
+    type Error;
+
+    /// Draws a pixel on the display.
+    fn draw_pixel(&mut self, item: drawable::Pixel<C>) -> Result<(), Self::Error>;
+
+    /// Draws an object from an iterator over its pixels.
+    fn draw_iter<T>(&mut self, item: T) -> Result<(), Self::Error>
+    where
+        T: IntoIterator<Item = drawable::Pixel<C>>,
+    {
+        for pixel in item {
+            self.draw_pixel(pixel)?;
+        }
+
+        Ok(())
+    }
+
+    /// Returns the dimensions of the `DrawTarget` in pixels.
+    fn size(&self) -> Size;
+
+    /// Clears the display with the supplied color.
+    ///
+    /// This default implementation can be replaced if the implementing driver provides an
+    /// accelerated clearing method.
+    fn clear(&mut self, color: C) -> Result<(), Self::Error>
+    where
+        Self: Sized,
+    {
+        primitives::Rectangle::new(Point::zero(), Point::zero() + self.size())
+            .into_styled(PrimitiveStyle::with_fill(color))
+            .draw(self)
+    }
+
+    /// Draws a styled line primitive.
+    ///
+    /// This default trait method can be overridden if a display provides hardware-accelerated
+    /// methods for drawing lines.
+    ///
+    /// # Caution
+    ///
+    /// This method should not be called directly from application code. It is used to define the
+    /// internals of the [`draw`] method used for the [`Styled`] [`Line`] primitive. To draw a line,
+    /// call [`draw`] on a [`Line`] primitive object.
```

Comment:
Better, but not quite there yet. `draw` cannot be called on `Line`, because only `Styled<Line>` has a draw method.